pandas/io/pickle.py

```diff
@@ -94,7 +94,14 @@ def to_pickle(
         is_text=False,
         storage_options=storage_options,
     ) as handles:
-        pickle.dump(obj, handles.handle, protocol=protocol)  # type: ignore[arg-type]
+        if handles.compression["method"] in (None, "gzip"):
+            pickle.dump(
```

Comment:
can we just always do this? (the new one)

Reply:
we can always do `handles.handle.write(...)` (<1.2) but this does not take advantage of pickle 5's zero-copy mechanism. Using `pickle.dump(obj, handles.handle)` (1.2.0 and 1.2.1) would be the preferable solution as it can take advantage of the zero-copy mechanism. I think the preferred solution is causing two issues together with compression:
1) it seems to call `write` multiple calls (that is only(?) an issue for zip: it creates multiple files instead of one file: but we have a write buffer to avoid this issue (already fixed in 1.2.1))
2) with pickle 5, it returns [PickleBuffer](https://www.python.org/dev/peps/pep-0574/) which doesn't seem to be compatible with `len` (the `TypeError` in `lzma` in the linked issue); that is probably a python bug (either in lzma/bz or pickle).

If (2) gets fixed upstream, we will be able to use `pickle.dump(obj, handles.handle)` in all cases. I think that the if/else is maintainable and provides users with more of pickle 5's features. An alternative is to have a write buffer in `get_handle` for zip/lzma/bz2 (I think that would also simplify some code in `to_stata`).

Reply:
ok can you add comments to this effect on why doing it this way for future readers. (and refernce the pickle 5 issue as well)